src/verifai/samplers/multi_armed_bandit.py

```diff
@@ -0,0 +1,203 @@
+        # every restart_every samples.
+        if self.is_multi:
+            if self.monitor is not None and self.monitor.linearize and t % self.restart_every == 0:
+                self.monitor._linearize()
+            self.update_dist_from_multi(sample, info, rho)
+            return
+        for i, b in enumerate(info):
+            self.counts[i][b] += 1.
+            if rho < self.thres:
+                self.errors[i][b] += 1.
+        # print(self.errors / self.counts)
+
+    def set_graph(self, graph):
+        self.priority_graph = graph
+        try:
+            self.thres = [self.thres] * graph.number_of_nodes()
+        except Exception as e:
+            print(e)
+            assert len(self.thres) == graph.number_of_nodes(), 'Must have as many thresholds as graph nodes'
+        self.num_properties = graph.number_of_nodes()
+        self.is_multi = True
+
+    # is rho1 better than rho2?
+    # partial pre-ordering on objective functions, so it is possible that:
+    # is_better_counterexample(rho1, rho2)
+    # and is_better_counterxample(rho2, rho1) both return False
+    def is_better_counterexample(self, ce1, ce2):
+        if ce2 is None:
+            return True
+        all_same = True
+        already_better = [False] * self.num_properties
+        for node in nx.dfs_preorder_nodes(self.priority_graph):
+            if already_better[node]:
+                continue
+            b1 = ce1[node]
+            b2 = ce2[node]
+            all_same = all_same and b1 == b2
+            if b2 and not b1:
+                return False
+            if b1 and not b2:
+                already_better[node] = True
+                for subnode in nx.descendants(self.priority_graph, node):
+                    already_better[subnode] = True
+        return not all_same
+
+    def _get_total_counterexamples(self):
+        return sum(self.counterexamples.values())
+    
+    @property
+    def counterexample_values(self):
+        return [ce in self.counterexamples for ce in self.rho_values]
+
+    def _add_to_running(self, ce):
+        if ce in self.counterexamples:
+            return True
+        to_remove = set()
+        # if there is already a better counterexample, don't add this.
+        if len(self.counterexamples) > 0:
+            for other_ce in self.counterexamples:
+                if self.is_better_counterexample(other_ce, ce):
+                    return False
+        # remove all worse counterexamples than this.
+        for other_ce in self.counterexamples:
+            if self.is_better_counterexample(ce, other_ce):
+                to_remove.add(other_ce)
+        for other_ce in to_remove:
+            del self.counterexamples[other_ce]
+        self.counterexamples[ce] = np.array([np.zeros(int(b)) for b in self.buckets])
+        return True
+    
+    def update_dist_from_multi(self, sample, info, rho):
+        try:
+            iter(rho)
+        except:
```

Comment:
Should catch a specific `Exception` here (or use `isinstance` instead) -- currently this will catch everything, including `KeyboardInterrupt`!